server/src/main/java/org/elasticsearch/cluster/routing/BatchedRerouteService.java

```diff
@@ -105,7 +107,10 @@ public ClusterState execute(ClusterState currentState) {
                             pendingRerouteListeners = null;
                         }
                     }
-                    if (currentListenersArePending) {
+                    if (currentState.blocks().hasGlobalBlock(STATE_NOT_RECOVERED_BLOCK)) {
+                        logger.trace("skipping reroute prior to state recovery [{}]", reason);
+                        return currentState;
```

Comment:
I think we are changing the contract slightly here in that a reroute may now result in the listener being invoked with a cluster-state that was not rerouted. Most of those listeners just do logging. One exception is the `DiskThresholdMonitor` which seems unproblematic wrt. this. But it still seems like a possibly confusing semantics for future users of the listener callback from the reroute service.

I am not sure I would prefer to defer invoking the listener until the block is removed, but I wonder what your thoughts are on this. I think it at least requires an explanation in the javadoc for the `RerouteService` method.

Reply:
Thanks, I now better understand your thinking here. I wonder if we should (not necessarily in this PR) add an assert to cluster state building that it cannot have a non-empty routing table with a `STATE_NOT_RECOVERED_BLOCK` block?

Reply:
Sure, I think https://github.com/elastic/elasticsearch/pull/83133 should do the trick, let's see what CI says.